In [3]:
import pandas as pd
import pandas
from sklearn.cross_validation import train_test_split
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt

In [15]:
alldata = pd.read_csv('finalmaster-ratios.csv')
alldata.head()

# Purchases  B01001001   B01001002  B01001003  B01001004  B01001005  \
0           22     206252  469.226965  31.432422  35.219052  33.628765   
1            7      61399  486.538869  22.899396  21.531295  27.036271   
2            3      73170  489.859232  28.905289  36.271696  28.235616   
3           94     251724  505.585483  32.054949  31.757004  28.102207   
4            0      37382  495.586111  25.413301  29.318924  26.162324   

   B01001006  B01001007  B01001008  B01001009    ...      B19001008  \
0  20.121017  12.610787   6.734480   6.225394    ...      49.409690   
1  16.808091  28.355511  18.192479  13.534422    ...      59.231680   
2  21.566216  12.218122   7.243406   7.380074    ...      63.996993   
3  18.651380  12.080692   7.035483   7.686991    ...      54.790900   
4  19.260607  12.893906   6.580707   7.062222    ...      58.883378   

   B19001009  B19001010  B19001011   B19001012   B19001013  B19001014  \
0  53.306757  42.318307  83.167229   89.249208  102.141470  52.872330   
1  50.093078  40.700626  92.612963  117.363344  113.344051  75.774243   
2  47.322923  42.505211  70.420610   90.033143   98.677692  54.703249   
3  48.681562  43.873381  84.717507  112.204444  127.137252  83.019904   
4  51.761414  47.310187  81.902582   93.793717  130.103014  71.982704   

   B19001015  B19001016  B19001017  
0  36.440765  23.446284  21.197485  
1  33.000508  33.169741  24.792689  
2  20.125056  11.890525  16.537397  
3  43.731067  38.851729  40.427349  
4  36.118530  31.603714  19.648989  

[5 rows x 190 columns]

In [7]:
allvariablenames = list(alldata.columns.values)

In [156]:
after_drop = alldata.drop(alldata.columns[0:8],axis=1)
after_drop.head()
allvariablenames = list(after_drop)
allvariablenames

['B01001008',
 'B01001009',
 'B01001010',
 'B01001011',
 'B01001012',
 'B01001013',
 'B01001014',
 'B01001015',
 'B01001016',
 'B01001017',
 'B01001018',
 'B01001019',
 'B01001020',
 'B01001021',
 'B01001022',
 'B01001023',
 'B01001024',
 'B01001025',
 'B01001026',
 'B01001027',
 'B01001028',
 'B01001029',
 'B01001030',
 'B01001031',
 'B01001032',
 'B01001033',
 'B01001034',
 'B01001035',
 'B01001036',
 'B01001037',
 'B01001038',
 'B01001039',
 'B01001040',
 'B01001041',
 'B01001042',
 'B01001043',
 'B01001044',
 'B01001045',
 'B01001046',
 'B01001047',
 'B01001048',
 'B01001049',
 'B02001001',
 'B02001002',
 'B02001003',
 'B02001004',
 'B02001005',
 'B02001006',
 'B02001007',
 'B02001008',
 'B02001009',
 'B02001010',
 'B12001001',
 'B12001002',
 'B12001003',
 'B12001004',
 'B12001005',
 'B12001006',
 'B12001007',
 'B12001008',
 'B12001009',
 'B12001010',
 'B12001011',
 'B12001012',
 'B12001013',
 'B12001014',
 'B12001015',
 'B12001016',
 'B12001017',
 'B12001018',
 'B12001019',
 'B130

In [139]:
predictors = after_drop
# from B01001008

len(predictors)

732

In [140]:
target = alldata['# Purchases']
len(target)

732

In [141]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)


In [142]:
from sklearn.linear_model import LassoLarsCV
from sklearn.datasets import make_regression
model = LassoLarsCV(precompute=False,cv=10).fit(pred_train,tar_train)


/Users/yifanli/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 5 iterations, i.e. alpha=1.496e+00, with an active set of 5 regressors, and the smallest cholesky pivot element being 4.215e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/Users/yifanli/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=1.098e+00, with an active set of 8 regressors, and the smallest cholesky pivot element being 4.215e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/Users/yifanli/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 10 iterations, i.e. alpha=7.329e-01, with an active set of 10 regressors,

/Users/yifanli/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 65 iterations, i.e. alpha=1.149e-01, with an active set of 47 regressors, and the smallest cholesky pivot element being 6.909e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/Users/yifanli/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 102 iterations, i.e. alpha=5.744e-02, with an active set of 80 regressors, and the smallest cholesky pivot element being 6.909e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/Users/yifanli/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 102 iterations, i.e. alpha=5.728e-02, with an active set of 80 regre

In [154]:
predictors_model=pd.DataFrame(allvariablenames)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)
        
        


['B01001036' 2.7861365955132507]
['B01001037' 0.9200572652790069]
['B01001038' 0.9459340522644333]
['B02001005' 0.39156809216155525]
['B13014026' 0.22056164158451835]
['B13014027' 0.05049787197081092]
['B19001017' 1.6062678580473928]


Question #1

first, we all assigning all variables name to predictors_model, and then get all the coeffecient out. The for loop will print out all coefficients, which is greater then 0

Question #2

['B01001014' 0.8557908775529921] Males aged 40 to 44 Years.
"In areas where there are more Males aged 40-44, we sell more 0.8557908775529921 unit Bobo Bars."

 ['B01001036' 2.505392496591849] Females aged 30 to 34 Years.
In areas where there are more Females aged 30-34, we sell more 2.505392496591849 unit Bobo Bars.

 ['B01001037' 0.8894214357013622] Females aged 35 to 39 Years.
In areas where there are more Females aged 35-39, we sell more 0.8894214357013622 unit Bobo Bars.

 ['B01001038' 1.5315839680821497] Females aged 40 to 44 Years.
In areas where there are more Females aged 40-44, we sell more 1.5315839680821497 unit Bobo Bars.

 ['B02001005' 0.4125408937426837] Asian Alone
In areas where there are asian alone, we sell 0.4125408937426837 unit Bobo Bars.

 ['B13014026' 0.4800240326923769] Women 15 to 50 Years Who Had a Birth in the Past 12 Months with Bachelor's Degree
In areas where there are more females aged 15-50 years Who Had a Birth in the Past 12 Months with Bachelor's Degree, 
we sell more 0.4800240326923769 unit Bobo Bars.

 ['B13014027' 0.6977454940063235] Women 15 to 50 Years Who Had a Birth in the Past 12 Months with Graduate or Professional Degree
In areas where there are more female aged 15 to 50 Years Who Had a Birth in the Past 12 Months with Graduate or Professional Degree, 
we sell more 0.6977454940063235 unit Bobo Bars.

 ['B13016001' 874922971.7249781] Women 15 to 50 Years Who Had a Birth in the Past 12 Months
In areas where there are more females aged 15-50 Years Who Had a Birth in the Past 12 Months, we sell more 874922971.7249781 unit Bobo Bars.

 ['B19001017' 1.4834465563617387] Household with income $200,000 or More
In areas where there are Household with income $200,000 or More, we sell more 1.4834465563617387 unit Bobo Bars.

Question #3

I would select B01001036 and B19001017 because they have higher coefficient 

In [152]:
from sklearn.metrics import mean_squared_error

train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
22528.486826258624


In [153]:
train_test = mean_squared_error(tar_test, model.predict(pred_test))
print ('training data MSE')
print(train_test)


training data MSE
41578.280293705764


Question #4 

In [159]:
#r squared
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.22266652028942102


In [160]:
rsquared_test=model.score(pred_test,tar_test)
print ('testing data R-square')
print(rsquared_test)

testing data R-square
0.17529294561525344


From the result above, we can observe that these two R-squares are different

Question #5

In [161]:
print("y interecept:")
print(model.intercept_)

y interecept:
2.758738710322305


Answer: This is not a great method to predict sales. Because census data cannot reflect the result very well. 
To be specific, training data MSE is huge, but R-square is very small. 

Question #6

The y intercept is 2.758738710322305, and the baseline sales number is 2.758738710322305.